File: llm_generated_code/ChatGPT/Python/Level1/Q9.py
def fractional_knapsack(weights, values, capacity):
    # Calculate the value-to-weight ratio for each item and store them as tuples
    items = [(values[i] / weights[i], weights[i], values[i]) for i in range(len(weights))]
    
    # Sort items by their value-to-weight ratio in descending order
    items.sort(reverse=True, key=lambda x: x[0])

    total_value = 0.0  # To keep track of the total value of items in the knapsack
    for ratio, weight, value in items:
        if capacity >= weight:
            # If the entire item can fit in the knapsack, take it whole
            total_value += value
            capacity -= weight
        else:
            # Take the fraction of the remaining capacity
            total_value += ratio * capacity
            break  # Knapsack is full, break out of the loop

    return total_value

# Example usage
weights = [10, 20, 30]
values = [60, 100, 120]
capacity = 50
print(f"Maximum value in the knapsack: {fractional_knapsack(weights, values, capacity)}")

File: llm_generated_code/ChatGPT/CPP/Level1/Q9.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Item {
    int weight;
    int value;
};

bool compare(Item a, Item b) {
    double r1 = (double)a.value / a.weight;
    double r2 = (double)b.value / b.weight;
    return r1 > r2;
}

double fractionalKnapsack(int W, vector<Item>& items) {
    // Sort items by value-to-weight ratio in descending order
    sort(items.begin(), items.end(), compare);

    double maxValue = 0.0;
    int currentWeight = 0;

    for (const auto& item : items) {
        if (currentWeight + item.weight <= W) {
            // If the item can be added in whole
            currentWeight += item.weight;
            maxValue += item.value;
        } else {
            // Take the fraction of the item that fits
            int remainingWeight = W - currentWeight;
            maxValue += item.value * ((double)remainingWeight / item.weight);
            break; // Knapsack is full
        }
    }

    return maxValue;
}

int main() {
    int W = 50; // Knapsack capacity
    vector<Item> items = {{10, 60}, {20, 100}, {30, 120}}; // {weight, value}

    double maxVal = fractionalKnapsack(W, items);
    cout << "Maximum value in Knapsack: " << maxVal << endl;

    return 0;
}

